util/src/main/scala/Money.scala

```diff
@@ -128,25 +202,248 @@ object Money {
+  def USD(amount: BigDecimal, precision: Option[Int] = None) = simpleValueMeantToBeUsedOnlyInTests(amount, "USD", precision)
+  def GBP(amount: BigDecimal, precision: Option[Int] = None) = simpleValueMeantToBeUsedOnlyInTests(amount, "GBP", precision)
+  def JPY(amount: BigDecimal, precision: Option[Int] = None) = simpleValueMeantToBeUsedOnlyInTests(amount, "JPY", precision)
+
+  val TypeName = "preciseAmount"
+
+  private def simpleValueMeantToBeUsedOnlyInTests(amount: BigDecimal, currencyCode: String, precision: Option[Int] = None): HighPrecisionMoney = {
+    val currency = Currency.getInstance(currencyCode)
+
+    val fractionDigits = precision.getOrElse(currency.getDefaultFractionDigits)
+
+    fromDecimalAmount(amount, fractionDigits, currency)(BigDecimal.RoundingMode.HALF_EVEN)
+  }
+
+  def roundToCents(amount: BigDecimal, currency: Currency)(implicit mode: RoundingMode): Long =
+    (amount.setScale(currency.getDefaultFractionDigits, mode) / centFactor(currency)).toLong
+
+  def sameScale(m1: HighPrecisionMoney, m2: HighPrecisionMoney): (BigDecimal, BigDecimal, Int) = {
+    val newFractionDigits = math.max(m1.fractionDigits, m2.fractionDigits)
+
+    def scale(m: HighPrecisionMoney, s: Int) =
+      if (m.fractionDigits < s) m.amount.setScale(s)
+      else if (m.fractionDigits == s) m.amount
+      else throw new IllegalStateException("Downscale is not allowed/expected at this point!")
+
+    (scale(m1, newFractionDigits), scale(m2, newFractionDigits), newFractionDigits)
+  }
+
+  def calc(m1: HighPrecisionMoney, m2: HighPrecisionMoney, fn: (BigDecimal, BigDecimal) ⇒ BigDecimal)(implicit mode: RoundingMode) = {
+    BaseMoney.requireSameCurrency(m1, m2)
+
+    val (a1, a2, fd) = sameScale(m1, m2)
+
+    fromDecimalAmount(fn(a1, a2), fd, m1.currency)
+  }
+
+  def factor(fractionDigits: Int) = BigDecimal(1) / BigDecimal(10).pow(fractionDigits)
+  def centFactor(currency: Currency) = factor(currency.getDefaultFractionDigits)
+
+  def fromDecimalAmount(amount: BigDecimal, fractionDigits: Int, currency: Currency)(implicit mode: RoundingMode) = {
+    val scaledAmount = amount.setScale(fractionDigits, mode)
+
+    HighPrecisionMoney(scaledAmount, fractionDigits, roundToCents(scaledAmount, currency), currency)
+  }
+
+  def fromCentAmount(centAmount: Long, fractionDigits: Int, currency: Currency) = {
+    val amount = BigDecimal(centAmount) * centFactor(currency)
+
+    HighPrecisionMoney(amount.setScale(fractionDigits, BigDecimal.RoundingMode.UNNECESSARY), fractionDigits, centAmount, currency)
+  }
+
+  def zero(fractionDigits: Int, currency: Currency) =
+    fromCentAmount(0L, fractionDigits, currency)
+
+  /* centAmount provides an escape hatch in cases where the default rounding mode is not applicable */
+  def fromPreciseAmount(preciseAmount: Long, fractionDigits: Int, currency: Currency, centAmount: Option[Long]) = {
```

Comment:
I don't really understand the comment above. At least, shouldn't `centAmount` default to `None`?

Reply:
At the moment `fromPreciseAmount` is only used in `FromJSON`-based deserialization and not intended for the normal use: `fromDecimalAmount`/`fromCentAmount` arshould be used instead. this is why I haven't added the default value, just wanted to make sure that `centAmount` is always explicitly propagated.

I'm still not 100% sure about the `centAmount` here though. The problem it tries to solve: `centAmount` is normally calculated based on rounding of the precise amount (for HP money class). It is calculated at the moment of `HPMoney` creation and saved as a field. The reason for this is to be able to propagate the appropriate rounding mode. This wouldn't be possible to do accurately if we calculate the centAmount when we serialize the value to JSON, since 
we don't have the appropriate rounding mode at this point (in `ToJSON`).

In general it works fine since in most place we can implicitly/explicitly propagate the rounding mode throughout the code. The only problematic part is `FromJSON`. In it we don't have any rounding mode available in order to calculate the `centAmount` for a deserialized HP money instance. So currently the strategy in this scenario is:

* Use the default rounding mode for the centAmount. I put a TODO here in order to discuss this approach and whether it makes sense: https://github.com/sphereio/sphere-scala-libs/pull/34/files#diff-9a30660394d2fcbf607bc8d29b74bd74R463
* In cases where API client would like to use different rounding mode for the `centAmount`, it can also add rounded `centAmount` value in the HP money JSON object. We will then use it and avoid rounding on our side. This is why I called it "an escape hatch" in the comment.

What do you think about this approach? Does this explanation helps?

Reply:
Yes, the explanation does help 👍 

Maybe you can expand the comment, I didn't get short version :)

For the TODO: IMO we could also always force the API client to provide the rounded centAmount in this case.